### PR TITLE
multi-network: fix data race on service update

### PR DESCRIPTION
**Pull request by howardjohn**
Detected by unit test -race flag

### CODE REVIEW

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 1]
```diff
@@ -404,8 +404,10 @@ func (c *Controller) deleteService(svc *model.Service) {
 	c.Lock()
 	delete(c.servicesMap, svc.Hostname)
 	delete(c.nodeSelectorsForServices, svc.Hostname)
+	c.networkManager.Lock()
```
Comment (keithmattix) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:407]: This feels kinda nasty to have to know about the details of networkManager's locking internals to use it. Does it make sense to extract any of this into a function that can handle the locking mechanism?

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:407]: This is also risky to have two locks

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:407]: Where did this panic happen

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:407]: I didn't make the two locks :stuck_out_tongue: I am just making sure we want to use them. If there is desire I can rewrite the network manage on krt.

The panic occurred here: https://github.com/istio/istio/blob/3b2c1db853269ca89d0ccf775a40b4d42b872fe8/pilot/pkg/serviceregistry/kube/controller/network.go#L233. Its a pretty straightforward "read during write" race condition.

Comment (keithmattix) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:407]: I'd be fine with merging this for now and create an issue to follow up with a rewrite/refactor. Might be good to help a new contributor get up to speed with krt

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:407]: I agree to merge first, the two lock can not rely on krt

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**